zenoh/src/net/routing/interceptor/downsampling.rs

```diff
@@ -0,0 +1,160 @@
+//
+// Copyright (c) 2024 ZettaScale Technology
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License 2.0 which is available at
+// http://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+//
+// Contributors:
+//   ZettaScale Zenoh Team, <zenoh@zettascale.tech>
+//
+
+//! ⚠️ WARNING ⚠️
+//!
+//! This module is intended for Zenoh's internal use.
+//!
+//! [Click here for Zenoh's documentation](../zenoh/index.html)
+
+use crate::net::routing::interceptor::*;
+use std::sync::{Arc, Mutex};
+use zenoh_config::DownsamplingItemConf;
+use zenoh_protocol::core::key_expr::OwnedKeyExpr;
+
+const RATELIMIT_STRATEGY: &str = "ratelimit";
+
+// EgressMsgDownsamplerRatelimit implements default ratelimit strategy, it has following
+// configuration parameters:
+//   threshold  - mandatory, indicates the minimum interval between messages.
+//                All messages arriving with a smaller interval will be skipped
+//   keyexprs   - optional, list of key expressions, only messages matching this expression will be processed,
+//                the rest will be passed as is.
+//   interfaces - optional, list of interfaces messages will be processed on,
+//                the rest will be passed as is.
+//
+// TODO: we can also add decimation strategy, with mandatory "factor" parameter
+
+pub(crate) struct EgressMsgDownsamplerRatelimit {
+    keyexprs: Option<Vec<OwnedKeyExpr>>,
+    threshold: std::time::Duration,
+    latest_message_timestamp: Arc<Mutex<std::time::Instant>>,
+}
+
+impl InterceptorTrait for EgressMsgDownsamplerRatelimit {
+    fn intercept(
+        &self,
+        ctx: RoutingContext<NetworkMessage>,
+    ) -> Option<RoutingContext<NetworkMessage>> {
+        if let Some(cfg_keyexprs) = self.keyexprs.as_ref() {
+            let mut matched = false;
+            if let Some(keyexpr) = ctx.full_key_expr() {
+                for cfg_keyexpr in cfg_keyexprs {
+                    if cfg_keyexpr.intersects(&keyexpr) {
+                        matched = true;
+                    }
+                }
+            }
+            if !matched {
+                return Some(ctx);
+            }
+        }
+
+        let timestamp = std::time::Instant::now();
+        let mut latest_message_timestamp = self.latest_message_timestamp.lock().unwrap();
+
+        if timestamp - *latest_message_timestamp >= self.threshold {
+            *latest_message_timestamp = timestamp;
+            log::debug!("Interceptor: Passed threshold, passing.");
```

Comment:
I'd move the log to `trace` level.

Reply:
Thanks, but it was there only for testing, I will remove it (my bad, forgot to leave TODO message)